src/photon/test/photon_tests.c

```diff
@@ -476,12 +515,77 @@ TEST task_multi_dependency_test(void) {
+  /* We start off with num_workers children processes, but no workers
+   * registered yet. */
+  ASSERT_EQ(utarray_len(photon->photon_state->child_pids), num_workers);
+  ASSERT_EQ(utarray_len(photon->photon_state->workers), 0);
+
+  /* Make sure that each worker connects to the photon scheduler. This for loop
+   * will hang if one of the workers does not connect. */
+  for (int i = 0; i < num_workers; ++i) {
+    new_client_connection(photon->loop, photon->photon_fd,
+                          (void *) photon->photon_state, 0);
+  }
+
+  /* After handling each worker's initial connection, we should now have all
+   * workers accounted for, but we haven't yet matched up process IDs with our
+   * children processes. */
+  ASSERT_EQ(utarray_len(photon->photon_state->child_pids), num_workers);
+  ASSERT_EQ(utarray_len(photon->photon_state->workers), num_workers);
+
+  /* Each worker should register its process ID. */
+  for (int i = 0; i < utarray_len(photon->photon_state->workers); ++i) {
+    local_scheduler_client *worker =
+        *(local_scheduler_client **) utarray_eltptr(
+            photon->photon_state->workers, i);
+    process_message(photon->photon_state->loop, worker->sock, worker, 0);
+  }
+  ASSERT_EQ(utarray_len(photon->photon_state->child_pids), 0);
+  ASSERT_EQ(utarray_len(photon->photon_state->workers), num_workers);
+
+  /* After killing a worker, its state is cleaned up. */
+  local_scheduler_client *worker = *(local_scheduler_client **) utarray_eltptr(
+      photon->photon_state->workers, 0);
+  kill_worker(worker, true);
+  ASSERT_EQ(utarray_len(photon->photon_state->child_pids), 0);
+  ASSERT_EQ(utarray_len(photon->photon_state->workers), num_workers - 1);
+
+  /* Start a worker after the local scheduler has been initialized. */
+  start_worker(photon->photon_state);
+  /* Accept the workers as clients to the plasma manager. */
+  int new_worker_fd = accept_client(photon->plasma_manager_fd);
+  /* The new worker should register its process ID. */
+  ASSERT_EQ(utarray_len(photon->photon_state->child_pids), 1);
+  ASSERT_EQ(utarray_len(photon->photon_state->workers), num_workers - 1);
+  /* Make sure the new worker connects to the photon scheduler. */
+  new_client_connection(photon->loop, photon->photon_fd,
+                        (void *) photon->photon_state, 0);
+  ASSERT_EQ(utarray_len(photon->photon_state->child_pids), 1);
+  ASSERT_EQ(utarray_len(photon->photon_state->workers), num_workers);
+  /* Make sure that the new worker registers its process ID. */
+  worker = *(local_scheduler_client **) utarray_eltptr(
+      photon->photon_state->workers, num_workers - 1);
+  process_message(photon->photon_state->loop, worker->sock, worker, 0);
+  ASSERT_EQ(utarray_len(photon->photon_state->child_pids), 0);
+  ASSERT_EQ(utarray_len(photon->photon_state->workers), num_workers);
+
+  /* Clean up. */
+  close(new_worker_fd);
+  destroy_photon_mock(photon);
+  PASS();
+}
+
```

Comment:
would be great to add a test that dynamically changes the number of workers : e.g.,
start 20 workers
repeat {
  kill 2 workers
  add 1 worker
} until num_workers > 0
This dynamic test will make sure we are not leaking worker state at runtime and that we're accounting for workers correctly (with asserts before/after each _worker operation )